### ISSUE TITLE
Elide redundant member events in `/messages` response

### ISSUE THREAD

**Issue by ShadowJonathan** [matrix-org/synapse#11960]
https://github.com/matrix-org/synapse/blob/68acb0a29dcb03a0ecbcebdb95e09c5999598f42/synapse/handlers/pagination.py#L525

This is not mandated explicitly by the spec;

> Rather than responses in the sequence sending duplicate membership events for these senders to the client, the server MAY assume that clients will remember membership events they have already been sent, and choose to skip sending membership events for members whose membership has not changed.

...however, this caused problems when conduit did not implement this similarly to synapse;

- https://gitlab.com/famedly/conduit/-/issues/233
- https://github.com/vector-im/element-android/issues/3417

JS-SDK (Cinny and Element Web), Element Android, and Hydrogen all assume this behaviour is a must.

I believe it'd help with more correct spec adoption greatly if redundant membership was indeed removed, so that current clients are more aware of this optional spec behaviour.

**Comment by erikjohnston** [comment]
We should do this on Synapse to help with spec compliance. However, before we can land the change in Synapse we'll need the clients to be fixed first.

We should talk to the client teams to make sure fixing those bugs is on their radar before we spend time on this.

**Comment by erikjohnston** [comment]
First step is to figure out what is required to actually fix these TODOs, note that there are likely other TODOs around lazy loading of members in the pagination and sync code.

**Comment by DMRobertson** [comment]
Possibly related: https://github.com/matrix-org/matrix-spec/issues/942

**Comment by ShadowJonathan** [comment]
https://github.com/matrix-org/synapse/issues/5393 appears to be related, and/or the original issue for this.

**Comment by richvdh** [comment]
duplicate of #5393
